### PR TITLE
MSC3783: Fixed base64 for SAS verification

### PR DESCRIPTION
**Pull request by uhoreg**
[Rendered](https://github.com/uhoreg/matrix-doc/blob/sas_fixed_base64/proposals/3783-fixed-base64-sas-verification.md)

Implementations:
- [JS SDK](https://github.com/matrix-org/matrix-js-sdk/pull/2320)
- [Polyjuice Client Test](https://gitlab.com/polyjuice/polyjuice_client_test/-/commit/1033a2803153f882c44e93c27e3aa9a27ad8b319)
  
image

[FCP tickyboxes](https://github.com/matrix-org/matrix-spec-proposals/pull/3783#issuecomment-1362168510)

### CODE REVIEW

**Comment by Liyoshinka** [comment]
please make an elixir homeserver will be better than dendrite

File: proposals/3783-fixed-base64-sas-verification.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# MSC3783: Fixed base64 for SAS verification
+
+libolm's original implementation for calculating the MAC for SAS-based device
```
Comment (richvdh) [inline proposals/3783-fixed-base64-sas-verification.md:3]: please could you link to the current spec, to help us understand what is changing?

Comment (uhoreg) [reply proposals/3783-fixed-base64-sas-verification.md:3]: The spec says the right thing, but the implementation in libolm was buggy.

**Comment by uhoreg** [comment]
@mscbot fcp merge

File: proposals/3783-fixed-base64-sas-verification.md (hunk 1 above)
Comment (richvdh) [reply proposals/3783-fixed-base64-sas-verification.md:3]: either way, links to the spec are really helpful.

I think we're talking about step 15 in https://spec.matrix.org/v1.5/client-server-api/#short-authentication-string-sas-verification ?

File: proposals/3783-fixed-base64-sas-verification.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+with SAS verification that uses the correct base64 encoding.  The current
+method will be deprecated.
+
+## Proposal
+
+A new message authentication code identifier `hkdf-hmac-sha256.v2` is
+introduced.  Clients that implement SAS verification are required to implement
+this method.  When the two clients that are verifying each other agree to use
+this method, the MAC is calculated in the same way as `hkdf-hmac-sha256`, but
+is encoded to base64 correctly.
```
Comment (richvdh) [inline proposals/3783-fixed-base64-sas-verification.md:22]: to be clear, `hkdf-hmac-sha256.v2` is used in the `message_authentication_code[s]` params of [`m.key.verification.start`](https://spec.matrix.org/v1.5/client-server-api/#mkeyverificationstartmsasv1)  and [`m.key.verification.accept`](https://spec.matrix.org/v1.5/client-server-api/#mkeyverificationaccept) ?

Comment (uhoreg) [reply proposals/3783-fixed-base64-sas-verification.md]: Yes.  Added some text to say so.

File: proposals/3783-fixed-base64-sas-verification.md (hunk 1 above)
Comment (uhoreg) [reply proposals/3783-fixed-base64-sas-verification.md]: Yes.  It's used in https://spec.matrix.org/v1.5/client-server-api/#mkeyverificationmac which I think is the better link, so I've added that

Comment (richvdh) [reply proposals/3783-fixed-base64-sas-verification.md]: thanks for the clarifications!

**Comment by uhoreg** [comment]
Spec PR: https://github.com/matrix-org/matrix-spec/pull/1412

**Comment by uhoreg** [comment]
merged! :tada:

**Pull request merged**